lib/src/alloc.c

```diff
@@ -0,0 +1,84 @@
+#include "alloc.h"
+
+#if defined(TREE_SITTER_ALLOCATION_TRACKING)
+
+void *ts_record_malloc(size_t);
+void *ts_record_calloc(size_t, size_t);
+void *ts_record_realloc(void *, size_t);
+void ts_record_free(void *);
+bool ts_toggle_allocation_recording(bool);
+
+static TSAllocator ts_tracking_allocator = {
+  .malloc = ts_record_malloc,
+  .calloc = ts_record_calloc,
+  .realloc = ts_record_realloc,
+  .free = ts_record_free,
+};
+
+TSAllocator *ts_allocator = &ts_tracking_allocator;
+
+#else
+
+#include <stdlib.h>
+
+bool ts_toggle_allocation_recording(bool value) {
+  (void)value;
+  return false;
+}
```

Comment:
Like I mentioned above, I think we'll want to simplify this system of tracking allocations once we add this feature, but I don't think you need to do it as part of this PR (unless you want to).

Reply:
I am not familiar with existing allocation tracking, so perhaps we can leave the refactoring in later PR.